### PR TITLE
Client-side login + reach router

### PR DESCRIPTION
**Pull request by carbonsoda**
Implemented reach router ([trello ticket](https://trello.com/c/tUYvEDlY/34-implementing-reach-router)) so I could start working on the history component, with help from Colin McDonald to get it setup.

The history link was placed on the navigation bar temporarily, but after some discussion about the mobile UI, I think it might be a better idea to place it at the top as well. In mobile view, this would be a collapsible "hamburger" style menu.

### CODE REVIEW

**Comment by carbonsoda** [comment]
[Merged in the client-login branch](https://github.com/carbonsoda/scribbler/pull/6) as it was a child of this branch and allowing it to be merged to main or even `frontend-v1` would then cause a merge conflict if I tried to include this older work.

File: app/.env.example [hunk 1]
```diff
@@ -0,0 +1,2 @@
+REACT_APP_AUTH0_DOMAIN=yourdomain.us.auth0.com
+REACT_APP_AUTH0_CLIENT_ID=yourclientid
```
Comment (mcab) [inline app/.env.example:2]: Due to `react-scripts`, these values will automatically get loaded during development / embedded during the build.

It's not quite self-explanatory how to set this up. Probably should put a comment in [app/README.md](https://github.com/carbonsoda/scribbler/blob/e4b3119bb2575fc34c6c8536f0d5ee1a6278b4c0/app/README.md) to indicate to follow Auth0's documentation on [Getting Started](https://auth0.com/docs/get-started) & [Registering a Single-Page Web Application](https://auth0.com/docs/applications/set-up-an-application/register-single-page-app) in order to get these values.

**Comment by mcab** [comment]
(whoops, did not read things too closely! It's actually [Reach Router](https://reach.tech/router/), and TIL I've learned something new 😃)

File: app/src/App.js [hunk 2]
```diff
@@ -1,19 +1,21 @@
...
     <div className="App">
       <Navbar />
-      <div className="main-page">
-        <DrawPanel />
-        <Sidebar />
-      </div>
+      <Router>
+        <Mainpage path="/" />
+        <History path="/history" />
+      </Router>
```
Comment (mcab) [inline app/src/App.js]: https://user-images.githubusercontent.com/4522043/116801220-852fad80-aabc-11eb-9f44-6968ff43689b.mp4

😎

File: app/src/App.js [hunk 3]
```diff
@@ -1,19 +1,21 @@
...
 export default function App() {
   return (
     <div className="App">
       <Navbar />
-      <div className="main-page">
-        <DrawPanel />
-        <Sidebar />
-      </div>
+      <Router>
+        <Mainpage path="/" />
```
Comment (mcab) [inline app/src/App.js]: nit: `Mainpage` does feel a bit wordy. You can simply call this `Main`.

File: app/src/components/Navbar.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+
+import { Link } from '@reach/router';
+
+import LoginItems from './navbar/LoginItems';
+
+export default function Navbar() {
+  return (
+    <div className="navbar">
```
Comment (mcab) [inline app/src/components/Navbar.js]: nit: `<nav>` over `<div className="navbar">` here. It's supported on all modern browsers (via [canIuse?](https://caniuse.com/?search=nav)).

File: app/src/components/navbar/LoginItems.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    if (isAuthenticated) {
+      logout({ returnTo: window.location.origin });
+    } else {
+      loginWithRedirect();
+    }
+  };
+  return (
+    <div className="login-items">
+      <LoginBtn isAuthenticated={isAuthenticated} loginUser={loginUser} />
+      <UserIcon user={user} />
```
Comment (mcab) [inline app/src/components/navbar/LoginItems.js:22]: nit: There's a scenario where the user's browser is still trying to determine its state of being logged in, or it's errored out.

One way to guard against this is to detect if the library is still loading, or it's errored out. This can be seen in [auth0/auth0-react#Getting Started](https://github.com/auth0/auth0-react#getting-started).

File: app/src/index.js [hunk 6]
```diff
@@ -1,12 +1,20 @@
...
 import App from './App';
 import reportWebVitals from './reportWebVitals';
 
 ReactDOM.render(
   <React.StrictMode>
-    <App />
+    <Auth0Provider
+      domain={process.env.REACT_APP_AUTH0_DOMAIN}
+      clientId={process.env.REACT_APP_AUTH0_CLIENT_ID}
+      redirectUri={window.location.origin}
```
Comment (mcab) [inline app/src/index.js:14]: nit: For now, this is okay. Ideally, this is changed to a value that you know where the application is being served from. In development, this is trickier, hence we rely on the current window's origin to set.

**Review by mcab — APPROVED** [review]
Overall, pretty good!

Anything with `nit:` is more my personal opinion on additional things to be done to address some edge cases. They're not necessary to address, but I'd be happy to discuss how to resolve them.

Of note, addressing the lack of authentication setup in the README would make it easier for someone to spin up on this project.
